base/notify.go

```diff
@@ -0,0 +1,247 @@
+package base
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"time"
+
+	"github.com/iter8-tools/iter8/base/log"
+)
+
+// notifyInputs is the input to the notify task
+type notifyInputs struct {
+	// URL is the URL of the notification hook
+	Url string `json:"url" yaml:"url"`
+
+	// Method is the HTTP method that needs to be used
+	Method string `json:"method,omitempty" yaml:"method,omitempty"`
+
+	// Params is the set of HTTP parameters that need to be sent
+	Params map[string]string `json:"params,omitempty" yaml:"params,omitempty"`
+
+	// Headers is the set of HTTP headers that need to be sent
+	Headers map[string]string `json:"headers,omitempty" yaml:"headers,omitempty"`
+
+	// URL is the URL of the request payload template that should be used
+	PayloadTemplateURL string `json:"payloadTemplateURL,omitempty" yaml:"payloadTemplateURL,omitempty"`
+
+	// SoftFailure indicates the task and experiment should not fail if the task
+	// cannot successfully send a request to the notification hook
+	SoftFailure bool `json:"softFailure" yaml:"softFailure"`
+}
+
+const (
+	// NotifyTaskName is the task name
+	NotifyTaskName = "notify"
+)
+
+// notifyTask sends notifications
+type notifyTask struct {
+	TaskMeta
+	With notifyInputs `json:"with" yaml:"with"`
+}
+
+// Report is the data that is given to the payload template
+type Report struct {
+	// Group string `json:"group" yaml:"group"`
+
+	// SLOs map[string]SLOReport `json:"SLOs" yaml:"SLOs"`
+
+	// Metrics map[string]MetricReport `json:"metrics" yaml:"metrics"`
+
+	// Timestamp is when the report was created
+	// For example: 2022-08-09 15:10:36.569745 -0400 EDT m=+12.599643189
+	TimeStamp string `json:"timeStamp" yaml:"timeStamp"`
+
+	// Completed is whether or not the experiment has completed
+	Completed bool `json:"completed" yaml:"completed"`
+
+	// NoTaskFailures is whether or not the experiment had any tasks that failed
+	NoTaskFailures bool `json:"noTaskFailures" yaml:"noTaskFailures"`
+
+	// NumTasks is the number of tasks in the experiment
+	NumTasks int `json:"numTasks" yaml:"numTasks"`
+
+	// NumCompletedTasks is the number of completed tasks in the experiment
+	NumCompletedTasks int `json:"numCompletedTasks" yaml:"numCompletedTasks"`
+
+	// NumLoops is the current loop of the experiment
+	NumLoops int `json:"numLoops" yaml:"numLoops"`
+
+	// Experiment is the experiment struct
+	Experiment *Experiment `json:"experiment" yaml:"experiment"`
+}
+
+// NotifyPayloadTemplateValues contains the report as well as the report in some other formats
+type NotifyPayloadTemplateValues struct {
```

Comment:
I wonder if this type is required at all. https://github.com/iter8-tools/iter8/blob/50f45970a5ccad4f616df80f8aba7d40999ba6c4/base/custom_metrics.go#L221 -- already includes Sprig's (Helm's) [type conversion template functions](https://helm.sh/docs/chart_template_guide/function_list/#type-conversion-functions)... and this in turn includes JSON conversion (without and without escaping).

Reply:
I see... I will give this a try and report back.